src/Elastic.Apm/Metrics/MetricsProvider/SystemTotalCpuProvider.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Elastic.Apm.Api;
+using Elastic.Apm.Logging;
+
+namespace Elastic.Apm.Metrics.MetricsProvider
+{
+	public class SystemTotalCpuProvider : IMetricsProvider
+	{
+		private const string SystemCpuTotalPct = "system.cpu.total.norm.pct";
+		private readonly IApmLogger _logger;
+
+		public SystemTotalCpuProvider(IApmLogger logger) => _logger = logger.Scoped(nameof(SystemTotalCpuProvider));
+
+		private bool _firstTotal = true;
+
+		private bool _getTotalProcessorTimeFailedLog;
+		private TimeSpan _lastCurrentTotalProcessCpuTime;
+		private DateTime _lastTotalTick;
+		private Version _processAssemblyVersion;
+		private PerformanceCounter _processorTimePerfCounter;
+
+		public int ConsecutiveNumberOfFailedReads { get; set; }
+		public string NameInLogs => "total system CPU time";
+
+		public IEnumerable<Sample> GetValue()
+		{
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+			{
+				if (_processorTimePerfCounter == null)
+					_processorTimePerfCounter = new PerformanceCounter("Processor", "% Processor Time", "_Total");
+
+				var val = _processorTimePerfCounter.NextValue();
+
+				return new List<Sample> { new Sample(SystemCpuTotalPct, (double)val / 100) };
+			}
+
+			//The x-plat implementation is ~18x slower than perf. counters on Windows
+			//Therefore this is only a fallback in case of non-Windows OSs
+			var timespan = DateTime.UtcNow;
+			TimeSpan cpuUsage;
+
+			foreach (var proc in Process.GetProcesses())
+			{
+				try
+				{
+					cpuUsage += proc.TotalProcessorTime;
+				}
+				catch (Exception)
+				{
+					//can happen if the current process does not have access to `proc`.
+					if (_getTotalProcessorTimeFailedLog) continue;
+
+					_getTotalProcessorTimeFailedLog = true;
+					_logger.Info()
+						?.Log(
+							"The overall CPU usage reported by the agent may be inaccurate. This is because the agent was unable to access the CPU usage of some other processes");
+				}
+			}
+
+			var currentTimeSpan = DateTime.UtcNow;
+
+			if (!_firstTotal)
+			{
+				if (_processAssemblyVersion == null)
+					_processAssemblyVersion = typeof(Process).Assembly.GetName().Version;
+
+				double cpuUsedMs;
+
+				//workaround for a CoreFx bug. See: https://github.com/dotnet/corefx/issues/37614#issuecomment-492489373
+				if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX) && _processAssemblyVersion < new Version(4, 3, 0))
+					cpuUsedMs = (cpuUsage - _lastCurrentTotalProcessCpuTime).TotalMilliseconds / 100;
+				else
+					cpuUsedMs = (cpuUsage - _lastCurrentTotalProcessCpuTime).TotalMilliseconds;
```

Comment:
The sum of CPU time from the previous call to `GetValue` is possibly over set of different processes. You can easily get negative value in `cpuUsage - _lastCurrentTotalProcessCpuTime`.

Reply:
Handled, we just throw away those values - only in the fallback implementation, on Windows and Linux this is not used.

Reply:
Negative values is just an example proving the approach as it is now is incorrect - just discarding negative values doesn't fix the algorithm. Now you can get arbitrarily wrong positive values. Of course the probability of it happening on desktop is lower than on server because on server with microservices/lamdas/functions processes for various runtimes will be spun in and out of existence very frequently.

One way to fix it is to maintain CPU times per PID and discards CPU times from the previous step for processes that don't exist anymore.

Reply:
Like I said in my [other comment](https://github.com/elastic/apm-agent-dotnet/pull/248#discussion_r291172885) I'm totally aware that this algorithm is not perfect - but it only runs as a fallback on non-Win and non-Linux OSs and I tested it on macOS, and it seems to be useful, plus we indicate in a log that this may be inaccurate.

I thought having this as a fallback is good for users on those rare OSs. But just on this part we can spend days, and that would not be sufficient. I would have kept this, but we should reach an agreement on this, so I'll simply remove this part.

Reply:
So what is your decision regarding fallback algorithm in this PR: fix it or remove it?

Reply:
Remove, like I said [here](https://github.com/elastic/apm-agent-dotnet/pull/248#discussion_r291304445).  - Still think it's a bad decision - accepting it for progress over perfection.

Reply:
If you prefer we can go with current implementation and open a technical debt issue to find a better algorithm for macOS.

Reply:
Is it even worth including for macOS? Does anyone run production applications on macOS? I'd personally be inclined to leave it out until requested, since macOS isn't really a server platform, and just report this metrics for Linux and Windows.

FWIW, you can probably get CPU info on macOS using P/Invoke: https://github.com/elastic/go-sysinfo/blob/master/providers/darwin/syscall_darwin_amd64.go#L175

Reply:
> Does anyone run production applications on macOS?

We have some data on this. In our survey 0,2% uses it as prod server, but 15,3% uses it for development, and my thinking was that first people will test the agent on a dev machine, so if we leave this out, for 15,3% we won't show CPU data in their dev env., furthermore unfortunately for free and total memory we atm only have Windows and Linux implementation, meaning in case of macOS both default APM UI charts would be empty if we also skip this one.

[That go code](https://github.com/elastic/go-sysinfo/blob/master/providers/darwin/syscall_darwin_amd64.go#L175) is very promising, we definitely should try to port it to .NET. 

I leave this part out for now since I seem to be the only one excited about it. We can revisit this later.